networking: ensure the netns directory is mounted

JulienBalestra opened the pull request:
### Context
Coming from Kubernetes issue [48427](https://github.com/kubernetes/kubernetes/issues/48427).
Initially I was thinking of doing it in the kubelet rkt code but as @squeed suggested, it's simpler and less hacky to do it in rkt codebase.

### TL;DR:
This PR ensure the `/run/netns` directory is always mounted once as tmpfs.
The `ip netns add <some-id>` command has this behavior by default. The kubelet calls this command for every new `hostNetwork: false` pods.

A mount of `/run/netns` over any existing netns files like `/run/netns/cni-30b37e42-157e-9796-dd74-ba4a3689bbde` cause issues with them.

### Warning
Upgrading rkt with this fix needs a full drain of any running rkt-netns containers on the node.

squeed commented:
I did some exploration, and sadly it's a bug in `iproute2`, not rkt. I'll outline the bug, the fix, and how we might be able to work around it.

`iproute2` doesn't actually do a `mount -t tmpfs ...`. Instead, it wants to make `/run/netns` a mountpoint with shared propagation so that an instance of `iproute2` inside a container can manage the host. This is a good idea. The basic sequence of actions it does are:

1. `mkdir -p /run/netns`
2. `mount --bind /run/netns /run/netns`
3. `mount --make-rshared /run/netns`

However, there is a bug in step 2. The bind mount keeps any already-existing files in `/run/netns`, but doesn't actually copy any mounts in `/run/netns`. Since network namespaces are recorded as bind mounts, this copies them as files but doesn't preserve the actual ns. The fix is for `iproute2` to do `mount --rbind`

`iproute2` skips step 2 if `/run/netns` is already a mountpoint. So, rkt should mirror the [logic in iproute`](https://kernel.googlesource.com/pub/scm/linux/kernel/git/shemminger/iproute2/+/master/ip/ipnetns.c#624), but with the bind mount an `rbind`.

Make sense?

JulienBalestra commented:
I made the changes suggested to _"mirror the logic in iproute"_.

I improved the error message returned by `syscall.Mount`. 
In case of failure, it's too short to be clear in the logs.

Example errno EINVAL:
`stage1: networking: failed to setup network: invalid argument`

lucab commented:
Everything up to here looks fine to me. I've milestoned this for the next release, but I'd like to wait and see whether we get some feedback on the iproute2 patch.

lucab commented on networking/podenv.go:
```diff
@@ -101,6 +101,31 @@ func (e *podEnv) loadNets() ([]activeNet, error) {
+		if err != nil {
+			return fmt.Errorf("mount --rbind %s %s failed: %q", mountNetnsDirectory, mountNetnsDirectory, err)
+		}
+		return e.mountNetnsDirectory(true)
```
Can we get rid of this recursive call here? It seems to be that if we reach this point we then just try one more `mount(rshared)` or fail hard.

JulienBalestra replied:
Ok, in this case I have no reason to leave the mkdir -p outside this method too.

lucab commented on networking/networking.go:
```diff
@@ -83,8 +84,18 @@ func Setup(podRoot string, podID types.UUID, fps []commonnet.ForwardedPort, netL
+		return nil, err
+	}
+
+	err = n.mountNetnsDirectory(false)
```
We were not doing this before, and perhaps if it doesn't succeed we don't need to hard-fail. Should we perhaps just log an error if the mount fails here and keep going? @squeed?

JulienBalestra replied:
The container will start well but we cannot guarantee the lifecycle of its network if a third-party software like `ip netns add` is triggered after its starts. So I'm definitely for a hard-fail.

JulienBalestra commented:
@lucab I removed the recursive call and moved the `mkdir` in the method.
Do we have a feedback for the iproute patch ? 
While I am here, do you know how long it could takes for that merged patch to be propagated into CoreOS container linux stable channel ?

lucab commented on networking/podenv.go:
```diff
@@ -101,6 +101,41 @@ func (e *podEnv) loadNets() ([]activeNet, error) {
+	err = syscall.Mount("", mountNetnsDirectory, "none", syscall.MS_SHARED|syscall.MS_REC, "")
+	if err != nil {
+		// Fail unless we need to make the mount point
+		if err.Error() != EINVAL {
```
You can simply use `syscall.EINVAL` here and get rid of your custom constant.

lucab commented on networking/podenv.go:
```diff
@@ -101,6 +101,41 @@ func (e *podEnv) loadNets() ([]activeNet, error) {
+	if err != nil {
+		// Fail unless we need to make the mount point
+		if err.Error() != EINVAL {
+			return fmt.Errorf("mount --bind %s failed: %q", mountNetnsDirectory, err)
```
The error message should start with `mount --make-rshared ...`.

lucab commented on networking/podenv.go:
```diff
@@ -101,6 +101,41 @@ func (e *podEnv) loadNets() ([]activeNet, error) {
+		// Remount after the Upgrade
+		err = syscall.Mount("", mountNetnsDirectory, "none", syscall.MS_SHARED|syscall.MS_REC, "")
+		if err != nil {
+			return fmt.Errorf("mount --bind %s failed: %q", mountNetnsDirectory, err)
```
The error message should start with `mount --make-rshared ...`.

squeed commented:
Tested myself, works great!
